### PR TITLE
[DebugInfo][DWARF] Set is_stmt on first non-line-0 instruction in BB

### LINKED ISSUE #104695: Missing is_stmt flag for 'else if' conditions
Originally from https://sourceware.org/bugzilla/show_bug.cgi?id=31896.

Sometimes if conditions don't get is_stmt flag and gdb skips them during stepping.

To reproduce, compile test.cpp:

```
 1  #include <iostream>
 2
 3  bool predicate(int v) {
 4      return v == 7;
 5  }
 6
 7  void f(int v) {
 8      if (v) { // break here
 9          std::cout << "branch1\n";
10      }
11      else if(predicate(v)) {
12          std::cout << "branch2\n";
13      } else if(predicate(v+1)) {
14          std::cout << "branch3\n";
15      }
16  }
17
18  int main() {
19      f(0);
20      return 0;
21  }
```

`> clang++-18 -g -O0 test.cpp`

Try stepping in gdb:
```
> gdb a.out
(gdb) b test.cpp:8
Breakpoint 1 at 0x116b: file test.cpp, line 8.
(gdb) run
Breakpoint 1, f (v=0) at test.cpp:8
8           if (v) { // break here
(gdb) next
11          else if(predicate(v)) {
(gdb) next
16      }
(gdb) 
```

Notice that gdb didn't stop at the last branch condition on line 13.

```
> llvm-dwarfdump-18 --debug-line a.out
...
Address            Line   Column File   ISA Discriminator OpIndex Flags
------------------ ------ ------ ------ --- ------------- ------- -------------
0x0000000000001140      3      0      0   0             0       0  is_stmt
0x0000000000001147      4     14      0   0             0       0  is_stmt prologue_end
0x000000000000114e      4      5      0   0             0       0 
0x0000000000001153      4      5      0   0             0       0  epilogue_begin
0x0000000000001160      7      0      0   0             0       0  is_stmt
0x000000000000116b      8      9      0   0             0       0  is_stmt prologue_end
0x000000000000116f      8      9      0   0             0       0 
0x0000000000001175      9     19      0   0             0       0  is_stmt
0x0000000000001188     10      5      0   0             0       0  is_stmt
0x000000000000118d     11     23      0   0             0       0  is_stmt
0x0000000000001190     11     13      0   0             0       0 
0x0000000000001195     11     13      0   0             0       0 
0x00000000000011a2     12     19      0   0             0       0  is_stmt
0x00000000000011b5     13      5      0   0             0       0  is_stmt
0x00000000000011ba     13     25      0   0             0       0 
0x00000000000011bd     13     26      0   0             0       0 
0x00000000000011c0     13     15      0   0             0       0 
0x00000000000011c5     13     15      0   0             0       0 
0x00000000000011d2     14     19      0   0             0       0  is_stmt
0x00000000000011e5      0     19      0   0             0       0 
0x00000000000011ef     16      1      0   0             0       0  is_stmt epilogue_begin
0x0000000000001200     18      0      0   0             0       0  is_stmt
0x000000000000120f     19      5      0   0             0       0  is_stmt prologue_end
0x0000000000001216     20      5      0   0             0       0  is_stmt
0x0000000000001218     20      5      0   0             0       0  epilogue_begin
0x000000000000121e     20      5      0   0             0       0  end_sequence
```

For the branch on the line 11 both closing `}` of the previous block on line 10 and the `else if` get is_stmt flag.

The branch on the line 13 has only one instruction marked with the is_stmt flag. Gdb treats it as the last instruction of the previous block and doesn't stop there.

### PR DESCRIPTION
**Pull request by SLTozer**
Fixes: https://github.com/llvm/llvm-project/issues/104695

This patch adds the is_stmt flag to line table entries for the first instruction with a non-0 line location in each basic block, to ensure that it will be used for stepping even if the last instruction in the previous basic block had the same line number; this is important for cases where the new BB is reachable from BBs other than the preceding block.

### CODE REVIEW

**Comment by SLTozer** [comment]
This does result in some is_stmts being added that are unnecessary, i.e. cases where the last instruction in the current BB's sole predecessor has the same line number as the first instruction in the current BB. We could potentially include that in the analysis, but I think there's no way to do so without doing a more complex (though probably not too expensive) analysis in DwarfDebug; we also wouldn't be able to account for indirect branches.

I think in any case this patch is an improvement if we're considering such cases, as not only might the current BB not have the previous BB as its sole predecessor, but the previous BB might not even be a predecessor - making the is_stmt-elision completely incorrect.

**Review by pogo59 — COMMENTED** [review]
I like this. I should have considered BB transitions when I did the line-0 accommodation in the first place. See question about the new test.

File: llvm/test/DebugInfo/Generic/is_stmt-at-block-start.ll [hunk 1]
```diff
@@ -0,0 +1,38 @@
+;; Checks that when an instruction at the start of a BasicBlock has the same
+;; DebugLoc as the instruction at the end of the previous BasicBlock, we add
+;; is_stmt to the new line, to ensure that we still step on it if we arrive from
+;; a BasicBlock other than the immediately preceding one.
+
+; RUN: %llc_dwarf -O0 -filetype=obj < %s | llvm-dwarfdump --debug-line - | FileCheck %s
+
+; CHECK:      {{0x[0-9a-f]+}}     13      5 {{.+}} is_stmt
+; CHECK-NEXT: {{0x[0-9a-f]+}}     13     25 {{.+}} is_stmt
```
Comment (pogo59) [inline llvm/test/DebugInfo/Generic/is_stmt-at-block-start.ll:9]: ISTM the first check is for block if.then2, the second is for if.else4? Should there also be one for if.end8?

Comment (SLTozer) [reply llvm/test/DebugInfo/Generic/is_stmt-at-block-start.ll:9]: Not for this test specifically, as only these two instructions have non-0 line numbers (excluding 0x0, which has the function header).

Comment (pogo59) [reply llvm/test/DebugInfo/Generic/is_stmt-at-block-start.ll:9]: Oh right, `!dbg` isn't inherited the way it is for asm.
But, you could use the same `!dbg` for both, showing that it really is the BB that triggers it.

**Review by pogo59 — APPROVED** [review]
LGTM with a test suggestion.

**Review by jryans — APPROVED** [review]
Looks like a very sensible improvement, thanks for working on this! 😄

File: llvm/test/DebugInfo/Generic/is_stmt-at-block-start.ll (hunk 1 above)
Comment (SLTozer) [reply llvm/test/DebugInfo/Generic/is_stmt-at-block-start.ll:9]: Hm, trying to make this work has revealed another an issue - when the lines are actually identical, they will be merged into a single entry (whether this patch is in effect or not), which any jump to the branch would land in the middle of. I'm not sure whether that would result in a step in the debugger or not; it prevents the test from working, at the very least.

**Comment by SLTozer** [comment]
Had a free moment so updated this - the previous version (as noted in my response to @pogo59 above) would still merge adjacent DLs across BB boundaries if they were exactly identical, rather than differing in column number; I've changed this to now only merge adjacent DLs within a BB, which strictly reduces the cases where we'll merge identical DLs.

**Review by jryans — APPROVED** [review]
Seems like a good tweak, looks good overall to me.

**Review by pogo59 — COMMENTED** [review]
Smooth. Still LGTM.

**Comment by SLTozer** [comment]
Updated a couple more tests whose outputs have changed as of the latest functional change - all of these look "correct" to me, in that they're adding is_stmt to instructions at the start of a basic block, or in `empty-inline.mir`, adding a few new entries that prevent a line-0 instruction at the start of a basic block from having a non-0 fallthrough location, and adding a new is_stmt entry at the start of a block (similar to the case in the initial report).

**Comment by SLTozer** [comment]
I added the test to the `Generic/` test dir since it doesn't test target-specific behaviour, but it does rely on the basic block layout which varies between targets; added commit 616f7d3d4f6d9bea6f776e357c938847e522a681 to move the test into `X86/`, which should fix the buildbot errors.

**Comment by SLTozer** [comment]
Hm, another [buildbot error](https://lab.llvm.org/buildbot/#/builders/160/builds/4142) - this time `createMCCodeEmitter failed`, which seems like it will also be a trivial fix, but I'll revert for now.

**Comment by dwblaikie** [comment]
We're seeing a pretty significant size regression from this change in self-host builds of clang and some other internal binaries we track.

For an internal debug build (with Split DWARF, compressed debug info sections in object files but uncompressed in the linked binary) we saw a total 3.6% increase in the size of the clang binary, a 12.7% increase in the size of the .debug_line section in object files, and a 18.3% increase in the size of .debug_line in the linked executable.

Might be worth reverting this and reconsidering whether the growth is acceptable or whether there's a more targeted fix that can be made?

**Comment by SLTozer** [comment]
> Might be worth reverting this and reconsidering whether the growth is acceptable or whether there's a more targeted fix that can be made?

I have another patch up now that tries to be more cautious with its application of is_stmt: https://github.com/llvm/llvm-project/pull/108251

It currently needs a fix, and confirmation that it has suitable performance, but it should result in fewer is_stmts being applied. With that said, if it doesn't bring the numbers down that much, there is a reasonable question about correctness vs size - it might not be feasible to compute the minimum set of is_stmts that don't result in missing steps, so there might be either a tradeoff to make, or a flag to add.

**Comment by dwblaikie** [comment]
Could you put this behind a flag or revert this for now (to see if you can find a solution that's sufficiently cautious to lower the size cost to something more acceptable) - I think the growth is a bit big to take by default? (perhaps other people have other opinions, though, in which case I'm open to that/we can take the hit if that's the prevailing opinion)

**Comment by SLTozer** [comment]
> Could you put this behind a flag or revert this for now

I can revert it, at least until the next patch is ready to land. I've tried to investigate this locally to see if the situation is improved by the next patch along, but my local build isn't showing anywhere close to the same increase - I've got .debug_line size increasing by 1.3% as a result of this commit. When you say "Debug", does that mean the CMake "Debug" option, or a release build with debug info enabled?

**Pull request merged**